homeassistant/components/aurora_abb_powerone/sensor.py

```diff
@@ -6,58 +6,105 @@
-    """Set up the Aurora ABB PowerOne device."""
+def setup_platform(hass, config: ConfigEntry, add_entities, discovery_info=None):
+    """Set up single sensor based on configuration.yaml (DEPRECATED)."""
     devices = []
     comport = config[CONF_DEVICE]
-    address = config[CONF_ADDRESS]
-    name = config[CONF_NAME]
-
+    address = config.get(CONF_ADDRESS, DEFAULT_ADDRESS)
+    name = config.get(CONF_NAME, "Solar PV")
+    _LOGGER.warning(
+        "DEPRECATED: setting up %s via configuration.yaml will "
+        "soon be unsupported.  Please remove the entries from your "
+        "configuration.yaml file and the set up the integration via the UI",
+        DOMAIN,
+    )
     _LOGGER.debug("Intitialising com port=%s address=%s", comport, address)
     client = AuroraSerialClient(address, comport, parity="N", timeout=1)
 
-    devices.append(AuroraABBSolarPVMonitorSensor(client, name, "Power"))
+    config.data = {"device_name": name}
+    config.entry_id = "undefined_entry_id"
+    devices.append(AuroraSensor(client, config, "Power", "instantaneouspower"))
     add_entities(devices, True)
 
 
-class AuroraABBSolarPVMonitorSensor(Entity):
-    """Representation of a Sensor."""
+async def async_setup_entry(hass, config, async_add_entities) -> None:
+    """Set up aurora_abb_powerone sensor based on a config entry."""
+    entities = []
+    client = hass.data[DOMAIN][config.entry_id]["client"]
+    serialnum = config.data[ATTR_SERIAL_NUMBER]
+    for sensor in hass.data[DOMAIN][config.entry_id]["devices"][serialnum]["sensor"]:
+        if sensor["parameter"] == "temperature":
+            entities.append(AuroraSensor(client, config, sensor["name"], "temperature"))
+        elif sensor["parameter"] == "instantaneouspower":
+            entities.append(
+                AuroraSensor(client, config, sensor["name"], "instantaneouspower")
+            )
+        else:
+            _LOGGER.error("Unrecognised sensor parameter '%s'", sensor["parameter"])
+    _LOGGER.debug("async_setup_entry adding %d entities", len(entities))
+    async_add_entities(entities, True)
+
+
+class AuroraSensor(AuroraDevice):
+    """Representation of a Sensor on a Aurora ABB PowerOne Solar inverter."""
+
+    availableprev = True
 
-    def __init__(self, client, name, typename):
+    def __init__(self, client: AuroraSerialClient, config_entry, name, typename):
         """Initialize the sensor."""
-        self._name = f"{name} {typename}"
-        self.client = client
         self._state = None
+        super().__init__(client, config_entry)
+
+        if typename == "instantaneouspower":
+            self.type = typename
+            self.units = POWER_WATT
+            self._device_class = DEVICE_CLASS_POWER
+        elif typename == "temperature":
+            self.type = typename
+            self.units = TEMP_CELSIUS
+            self._device_class = DEVICE_CLASS_TEMPERATURE
+        else:
+            _LOGGER.warning("Unrecognised typename '%s'", typename)
```

Comment:
Raise an error.

Reply:
Done.  I used InvalidStateError because it seemed the most fitting. But if there is a better one please say.